### PR TITLE
Draft: Mimic upgrade

### PR DESCRIPTION
**Pull request by tkschmidt**
- README changes/updates
- CMAKE min version 2.6 -> 3.18
- use c++17
- add conan
- add catch2
- add missing (implicit) includes (due to c++ version change?)
- added missing using namespace std (everything was written like this I just added it since it was missing)
- add tests
- possible bug fix when multiple shuffle runs: You can set a parameter to add the fasta shuffled n times. The shuffling tries to find new sequences for each shuffle by saving the known sequences. This set was reset for each of the n iterations.
- N->L, Q->E, K->E "replacements" for known shuffles are removed. Background To avoid creating AA shuffles with too similar masses some amino acids are not saved directly. They are converted to a new AA and then saved as known sequence. I decided that the changes above are not needed. They are usuassly ~1 Da apart and should also cause different fragmentation pattern. For ITMS K and Q can not be seperated (~0.05 Da difference) but we still expect a different fragmentation. So in the end only I->L conversion is left on that code section
- fix a bug in generating random AA which accessed random memory
- add option to avoid mutating I/L into other amino acids as well as mutating other amino acids into IL
- random mutations don't add cleavage points: No mutation to K/R
- add option to infer AA frequencies from the fasta file (these frequencies are used for AA mutations. Frequencies will always be printed to console. Only Regular AminoAcids are used. X,B,J,U,... are ignored.
- fix bug where a shuffle range was too large so random memory values were mixed into the sequence
- general code improvments: no double empty init of strings, removed useless code, added some sanity checks, include c++ headers instead of their C variants, use stl random code generator, 
- bug fix: C-term is now preserved (before only N-term was preserved)

### CODE REVIEW

File: mimic.Rproj
Comment (MatthewThe) [inline mimic.Rproj:1]: Do we need this file?

File: src/Peptides.cpp [hunk 1]
```diff
@@ -16,230 +16,364 @@
...
 }
 
-bool Peptides::recurringPeptide(const string& pep, bool force) {
+bool Peptides::checkAndMarkUsedPeptide(const string& pep, bool force) {
   string checkPep=pep;
-  if (replaceI) {
-    for(string::iterator it=checkPep.begin();it!=checkPep.end();it++) {
-      if (*it=='I')
-        *it='L';
-      if (*it=='N')
```
Comment (MatthewThe) [inline src/Peptides.cpp:166]: Did you remove the other equivalences because they're not relevant for high-resolution data?

**Review by MatthewThe — COMMENTED** [review]
Looks good. Just two minor comments/questions.

File: mimic.Rproj
Comment (micgrab) [reply mimic.Rproj:1]: Hi,
no this one needs to be removed. Seem like it was missed in the `.gitignore` file. Is it ok to add this file also to the `.gitignore` file?

File: src/Peptides.cpp (hunk 1 above)
Comment (micgrab) [reply src/Peptides.cpp:166]: Correct. Do you want this change reverted or an option to turn it on or off? When you are fine with an option for it how should it be implemented.


## Bool arg: 
E.g. `--low-mass-accuracy` to turn on the NQK conversions. This raises the issue how replaceI and low-mass-accuracy are interacting. low-mass-accuracy without replaceI does not make sense and I don't want to check for sensible argument combinations.


## Switch with 3 levels
- 0: no change
- 1: I replace
- 2: NQK replace  

CLI name: `ambiguity-level`, CLI default would be 0. NQK replace would always imply the I replace (via a fallthrough). This options sounds better to me.

File: src/Peptides.cpp [hunk 2]
```diff
@@ -16,230 +16,364 @@
...
+      "seed",
+      "Set seed of the random number generator. Default = 1",
+      "int");
+  cmd.defineOption("P",
+      "prepend",
+      "Prepend the original fasta file to the output",
+      "",
+      TRUE_IF_SET);
+    cmd.defineOption("I",
+                     "replaceI",
```
Comment (micgrab) [inline src/Peptides.cpp:315]: When this option is not replaced by the proposed switch then `-` should be used as separator.
```suggestion
                     "replace-I",
```

**Pull request merged**